[NCCL] Join work clean up thread before aborting communicators

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55444 [NCCL] Join work clean up thread before aborting communicators**

Changes ~ProcessGroupNCCL so that we join work cleanup thread before aborting nccl communicators. This is because if we abort nccl communicators first on destruction, outstanding work objects in `workMetaList` can have exceptions set on them. Right now this doesn't trigger errors in nccl async error handling due to the terminated check, but it seems a bit cleaner to just join this thread first.

The main motivation is also to reduce log spam since we added some logging when an exception is set on `WorkNCCL`, but this unexpectedly resulted in a lot of false-positive errors being logged even after pg shutdown. An example is below:

```
I0406 18:30:27.361981 1567104 ProcessGroupNCCL.cpp:527] [Rank 1] NCCL watchdog thread terminated normally
I0406 18:30:27.364675 1567105 ProcessGroupNCCL.cpp:265] [Rank 1] found async exception when checking for NCCL errors: NCCL error: unhandled system error, NCCL version 2.
7.3
```

With this change, we no longer see these false positive logs.

Differential Revision: [D27613035](https://our.internmc.facebook.com/intern/diff/D27613035/)

rohan-varma committed:
[rfc] Add terminated check in workCleanupLoop

After ProcessGroupNCCl is destroyed, it is possible that `workCleanupLoop` is still running and and processing possible timed out work. Although, we don't do anything if there is timed out work since `handleNcclGuard` is gated by `terminateProcessGroup_` check. So instead of iterating through workMetaList, just clear it if we are shutting down. This also reduces log spam since we added some logging when an exception is set on `WorkNCCL`, but this unexpectedly resulted in a lot of false-positive errors being logged even after pg shutdown. An example is below:

```
I0406 18:30:27.361981 1567104 ProcessGroupNCCL.cpp:527] [Rank 1] NCCL watchdog thread terminated normally
I0406 18:30:27.364675 1567105 ProcessGroupNCCL.cpp:265] [Rank 1] found async exception when checking for NCCL errors: NCCL error: unhandled system error, NCCL version 2.
7.3
```

Differential Revision: [D27613035](https://our.internmc.facebook.com/intern/diff/D27613035/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -669,21 +670,25 @@ void ProcessGroupNCCL::workCleanupLoop() {
-          // workNCCL objects from work vector.
-          if (!terminateProcessGroup_.load()) {
-            work.handleNCCLGuard();
+      if (terminateProcessGroup_.load()) {
```
I'm not sure if these conditionals help since the variable could be set to true right after you do this check (same goes for line 682). Is a better fix joining the workCleanupThread before aborting the communicators in ProcessGroupNCCL destructor?

rohan-varma replied:
Ah ok, so the root cause of those false positive error logs are due to us aborting the communicators during destruction. Yes, this seems like a better fix

rohan-varma committed:
Update on "[NCCL] Join work clean up thread before aborting communicators"


Changes ~ProcessGroupNCCL so that we join work cleanup thread before aborting nccl communicators. This is because if we abort nccl communicators first on destruction, outstanding work objects in `workMetaList` can have exceptions set on them. Right now this doesn't trigger errors in nccl async error handling due to the terminated check, but it seems a bit cleaner to just join this thread first.

The main motivation is also to reduce log spam since we added some logging when an exception is set on `WorkNCCL`, but this unexpectedly resulted in a lot of false-positive errors being logged even after pg shutdown. An example is below:

```
I0406 18:30:27.361981 1567104 ProcessGroupNCCL.cpp:527] [Rank 1] NCCL watchdog thread terminated normally
I0406 18:30:27.364675 1567105 ProcessGroupNCCL.cpp:265] [Rank 1] found async exception when checking for NCCL errors: NCCL error: unhandled system error, NCCL version 2.
7.3
```

With this change, we no longer see these false positive logs.

Differential Revision: [D27613035](https://our.internmc.facebook.com/intern/diff/D27613035/)

[ghstack-poisoned]

osalpekar reviewed (approved):
Looks good to me!
